Hard gas limits browser form

danfinlay opened the pull request:
Fixes #1234 
Fixes #1233
Fixes #1230
Fixes #1155
Fixes #1221.

Normal browser form invalid messages don't work in an extension-popup context, so I've implemented some custom messages here.

Right now they stay up indefinitely, maybe we want them to disappear after a couple seconds.

Opening as a possible proposal.

Here I show both a single validation, and what it looks like if two things are outside the valid range.

![screen shot 2017-03-22 at 3 03 45 pm](https://cloud.githubusercontent.com/assets/542863/24223431/4c931b5e-0f13-11e7-97ea-519e3038d57f.png)
![screen shot 2017-03-22 at 3 03 33 pm](https://cloud.githubusercontent.com/assets/542863/24223432/4cab3d6a-0f13-11e7-9021-9fe07717c18b.png)

danfinlay committed:
Some progress

danfinlay committed:
Got basic validations working

FlySwatter committed:
Merge branch 'master' into i1234-HardGasLimits-BrowserForm

danfinlay committed:
Bump changelog

2-am-zzz committed:
Merge branch 'master' into i1234-HardGasLimits-BrowserForm

FlySwatter committed:
Merge branch 'master' into i1234-HardGasLimits-BrowserForm

Zanibas commented:
So these errors are overlayed over the current interface?

It's good, but the styling needs to be worked on for sure.

flyswatter commented:
Yeah they're absolutely positioned, so they just hover under the fields they refer to.

@VladTod would you maybe whip us up a better form validation message?

VladTod commented:
Hello,  @flyswatter and @Zanibas !
Shouldn't the user have the chance to change the gas amount for example? to correct the error/warning.
We need to find a way to show the errors and let people change the field while the error is visible.
Maybe we could display them in the bottom? after the 3 buttons? and maybe just have an "*" next to the wrong field.

Will design it soon and show you what i mean

flyswatter commented:
Maybe we should ship this first, style it later.

I'm open to all kinds of styling improvements, but I also want us to ship this ASAP, because we have new users every day who report having their entire accounts locked up because they don't have this feature.

The current way does let users edit the field.

Right now the message disappears once the value is correct.

We could *also* make it disappear after a time delay, or after being clicked.

I look forward to seeing what Vlad proposes, but I also want to get an MVP out the door soon.

2-am-zzz committed:
Merge branch 'master' into i1234-HardGasLimits-BrowserForm

danfinlay committed:
Combine pending-tx-details component into pending-tx-details

These were only separated originally so we could make the notification-based TX approval work, which provided its own buttons.

This two templates are logically highly coupled, and keeping them working while separate has been difficult at times, and has even required resorting to dubious practices, like using React's `refs` pattern.

This combines them into one fairly large component, but I think it's ok, we can still break this up into components, just not the separation that it had previously.

danfinlay committed:
Merge branch 'i1234-HardGasLimits-BrowserForm' of github.com:MetaMask/metamask-plugin into i1234-HardGasLimits-BrowserForm

flyswatter commented:
Now also Fixes #1155.

kumavis commented on ui/app/conf-tx.js:
```diff
@@ -172,47 +172,53 @@ ConfirmTxScreen.prototype.onTxChange = function (txData) {
 }
 
+ConfirmTxScreen.prototype.stopPropagation = function (event) {
+  if (event.stopPropagation) {
```
whats this for? do some events not have this?

flyswatter replied:
Correct.  Specifically, the `invalid` event emitted by forms, so I just made this function so we could handle it generically with our other events in this component.

2-am-zzz reviewed (changes requested):
The reset button isn't functioning right now.

danfinlay committed:
Fix some broken refs

danfinlay committed:
Override browser default validation message

danfinlay committed:
Prevent default for reset and reject buttons

danfinlay committed:
Fix initial gas price estimate

danfinlay committed:
Make reset button clear errors

danfinlay committed:
Fix a couple things

Sorry apparently the gas fixes weren't in the last commit, but are in this one.

As reported in previous commit, fixes a bug where initial estimate is not derived from the network.

Also fixes a bug where clicking "reset" does not clear our custom validation warnings.

danfinlay committed:
Fix initial tx fee estimation

danfinlay committed:
Make gas calculations on render more consistent

2-am-zzz committed:
Merge branch 'master' into i1234-HardGasLimits-BrowserForm

danfinlay committed:
Disable accept button when gas limit is too low

danfinlay committed:
Merge branch 'i1234-HardGasLimits-BrowserForm' of github.com:MetaMask/metamask-plugin into i1234-HardGasLimits-BrowserForm

danfinlay committed:
Remove unnecessary log

Commit:
Remove goHome action causing erratic UI behavior.

danfinlay committed:
Fix tx selecting bug

Commit:
Add personalMessages to function to calculate pending tx index.

VladTod commented:
Hello Guys, here are two variations for the error "pop-up"
Important thing is that user should be able to edit the fields with the error visible. Otherwise people will forget what the error is, or what the number in the error is. For example on gas price error, 20000000 is hard to input without seeing it, because it has many 0's.

I'm not quite sure this first version is even possible since it's broken in 2 parts.
![pop-uperror](https://cloud.githubusercontent.com/assets/13063928/24394455/bd22ba6a-13a4-11e7-85ab-f0b221637c2b.jpg)

And this is the second variation.
![01-pop-uperror](https://cloud.githubusercontent.com/assets/13063928/24394474/c7d874cc-13a4-11e7-8403-1b997c190a24.jpg)

Let me know what you think
